### PR TITLE
Chitter Challenge

### PR DESCRIPTION
**Pull request by MattGough**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -44,68 +27,14 @@ I want to see all peeps in reverse chronological order
...
 
-You can see your [test coverage](https://github.com/makersacademy/course/blob/master/pills/test_coverage.md) when you submit a pull request, and you can also get a summary locally by running:
-
-```
-$ coveralls report
-```
+![Imgur](http://i.imgur.com/EbAGonr.png)
+![Imgur](http://i.imgur.com/1YVsv3D.png)
+![Imgur](http://i.imgur.com/wPkLXrA.png)
+![Imgur](http://i.imgur.com/QjU86BW.png)
```
Comment (tcpickard94) [inline README.md]: Consider adding installation instructions so it can be run locally

File: app/views/index.erb [hunk 2]
```diff
@@ -0,0 +1,48 @@
+<! DOCTYPE html>
```
Comment (tcpickard94) [inline app/views/index.erb]: You can remove DOCTYPE, html tags and body. These are defined for you - in layout

File: Gemfile [hunk 3]
```diff
@@ -1,6 +1,16 @@
 source 'https://rubygems.org'
 ruby '2.2.3'
 
+gem 'sinatra'
+gem 'rspec-sinatra'
+gem 'capybara'
+gem 'data_mapper'
+gem 'dm-postgres-adapter'
+gem 'dm-migrations'
+gem 'database_cleaner'
```
Comment (tcpickard94) [inline Gemfile]: Consider adding some of these into your test group

File: spec/spec_helper.rb [hunk 4]
```diff
@@ -1,8 +1,41 @@
+ENV["RACK_ENV"] = "test"
+
+require 'capybara/rspec'
 require 'coveralls'
 require 'simplecov'
 
+#require './app/models/message'
+require './app/app'
+require './app/data_mapper_setup'
```
Comment (tcpickard94) [inline spec/spec_helper.rb]: This can be required within you are and pulled in above ^

File: app/data_mapper_setup.rb [hunk 5]
```diff
@@ -0,0 +1,9 @@
+require 'data_mapper'
+require 'dm-postgres-adapter'
+
+require_relative 'models/message'
+require_relative 'models/user'
+
+DataMapper.setup(:default, ENV['DATABASE_URL'] || "postgres://localhost/chitter_#{ENV['RACK_ENV']}" )
+DataMapper.finalize
+DataMapper.auto_upgrade!
```
Comment (tcpickard94) [inline app/data_mapper_setup.rb]: Consider extracting this to an individual rake file - https://github.com/makersacademy/chitter-challenge/blob/master/docs/review.md#ensure-rakefile-has-appropriate-tasks

File: spec/add_user_spec.rb [hunk 6]
```diff
@@ -0,0 +1,24 @@
+require 'web_helper'
+
+feature 'User can sign up' do
```
Comment (tcpickard94) [inline spec/add_user_spec.rb]: Consider adding feature tests into their own features/ file

File: spec/web_helper.rb [hunk 7]
```diff
@@ -0,0 +1,14 @@
+def sign_up(name: 'Steve',
```
Comment (tcpickard94) [inline spec/web_helper.rb]: Consider making this a module

File: app/app.rb [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    erb :index
+  end
+
+  get '/users/new' do
+    @user = User.new
+    erb :sign_up
+  end
+
+  post '/users' do
+    @user = User.create(name: params[:name],
```
Comment (tcpickard94) [inline app/app.rb]: Consider refactoring some of this into a class method on the user model - https://github.com/makersacademy/chitter-challenge/blob/master/docs/review.md#refactor-long-controller-methods

File: app/app.rb [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+  enable :sessions
+  #set :session_secret, 'super secret'
+
+  get '/' do
+    @messages = Message.all.reverse
+    erb :index
+  end
+
+  get '/users/new' do
```
Comment (tcpickard94) [inline app/app.rb]: Consider splitting routes into individual controller files - https://github.com/makersacademy/chitter-challenge/blob/master/docs/review.md#split-routes-into-separate-controller-files

File: app/views/index.erb [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+ <a href="users/new">Sign Up</a>
+ <br >
+ or
+ <br >
+ <a href='/sessions/new'>Sign In</a>
+  <% end %>
+<% if @messages != nil %>
+<%= @current_user %>
+
+<ul id='messages'>
```
Comment (tcpickard94) [inline app/views/index.erb]: This could be wrapped in an article element

File: app/views/index.erb [hunk 11]
```diff
@@ -0,0 +1,48 @@
+<! DOCTYPE html>
+<html>
+
+<h1> CHITTER </h1>
+
+<body>
+  <% if current_user %>
+    Welcome, <%= current_user.username %> <%= current_user.email %>
+    <form method='post' action='/sessions'>
```
Comment (tcpickard94) [inline app/views/index.erb]: Consider extracting this into a partial?

File: app/app.rb [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+
+   delete '/sessions' do
+     session[:user_id] = nil
+     flash.keep[:notice] = 'goodbye!'
+     redirect to '/'
+   end
+
+
+
+  helpers do
```
Comment (tcpickard94) [inline app/app.rb]: This can be extracted into it's own helper file and required accordingly

**Comment by mechacoach** [comment]
You had your chitter-challenge reviewed by **Tom Pickard**

### The good points are:

* All unnecessary components removed, Ruby style guide conventions followed, useflash.nowandflash.nextcorrectly: flash.now and flash.next used correctly
* Gemfile uses test groups
* Database Cleaner is set up correctly
* DataMapper set up in separate configuration file, ensurespechelper.rbisconfiguredcorrectly: Only test helpers (not view helpers) are pulled in to spec_helper.rb
* Feature tests organized correctly
* Spec 'Helpers' used appropriately
* Asset routes set correctly
* Used modular style
* Forms use correct markup
* Correct use of Semantic HTML elements

### You should consider the following improvements:

* Installation Instructions should be in README
* Ensure Rakefile has Appropriate Tasks
* Ensure spec_helper.rb is Configured Correctly
* Ensure Correct Location of Feature/Acceptance and Unit Tests
* Use flash.now and flash.next Correctly
* Refactor Long Controller Methods
* Split Routes into Separate Controller files
* Use Partials with Standard HTML Conventions
* Use View Helpers

### Additional comments:
Great work!

see https://github.com/makersacademy/chitter-challenge/blob/master/docs/review.md for more details

**Comment by silvabox** [comment]
Hi Matt, obviously you've still got quite a bit to complete, but what you've done so far looks good.

If you get time to come back it, think about introducing a layout file.  Also think about your route names:  `post '/signupinfo'` could just be `post '/signup'` (or `sign_up`) - the `info` is superfluous and it doesn't match the assoicated `get` route.

**Comment by MattGough** [comment]
@silvabox Hi Ben, thanks for the comments, I don't think I have a  ```post '/signupinfo'``` route though?

**Comment by silvabox** [comment]
Sorry Matt - I'm such a dufus - I've posted that on the wrong pull request!

**Comment by MattGough** [comment]
Haha no worries, had me a bit confused there!

**Pull request closed without merging**